### ISSUE TITLE
FMA regression on windows on master (CPU:9600KF)

### ISSUE THREAD

**Issue by N5N3** [JuliaLang/julia#43386]
The detection seems incorrect? (I think 9600KF has native FMA support?)
On 1.7.0
```julia
julia> versioninfo()
Julia Version 1.7.0
Commit 3bf9d17731 (2021-11-30 12:12 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-9600KF CPU @ 3.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, skylake)
Environment:
  JULIA_CUDA_VERSION = 11.4.0

julia> @code_native fma(1.0,1.0,1.0)
        .text
; ┌ @ floatfuncs.jl:347 within `fma`
        pushq   %rbp
        movq    %rsp, %rbp
; │┌ @ floatfuncs.jl:336 within `fma_llvm`
        vfmadd213sd     %xmm2, %xmm1, %xmm0     # xmm0 = (xmm1 * xmm0) + xmm2
; │└
        popq    %rbp
        retq
        nopl    (%rax,%rax)
```
on master 
```julia
julia> versioninfo()
Julia Version 1.8.0-DEV.1090
Commit d16f4806e9 (2021-12-01 10:54 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-9600KF CPU @ 3.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, skylake)
Environment:
  JULIA_CUDA_VERSION = 11.4.0

julia> @code_native fma(1.0,1.0,1.0)
        .text
        .file   "fma"
        .globl  julia_fma_552                   # -- Begin function julia_fma_552
        .p2align        4, 0x90
        .type   julia_fma_552,@function
julia_fma_552:                          # @julia_fma_552
; ┌ @ floatfuncs.jl:416 within `fma`
        .cfi_startproc
# %bb.0:                                # %L5
        pushq   %rbp
        .cfi_def_cfa_offset 16
        .cfi_offset %rbp, -16
        movq    %rsp, %rbp
        .cfi_def_cfa_register %rbp
        subq    $32, %rsp
        movabsq $j_fma_emulated_554, %rax
        callq   *%rax
        addq    $32, %rsp
        popq    %rbp
        retq
.Lfunc_end0:
        .size   julia_fma_552, .Lfunc_end0-julia_fma_552
        .cfi_endproc
; └
                                        # -- End function
        .section        ".note.GNU-stack","",@progbits
```

**Comment by DilumAluthge** [comment]
Duplicate of https://github.com/JuliaLang/julia/issues/43088, maybe?

cc: @oscardssmith

**Comment by oscardssmith** [comment]
Your CPU has FMA, but your operating system has a broken version of it that causes problems.

**Comment by oscardssmith** [comment]
We could consider letting windows users keep their broken version given the performance difference and the fact that it's not our fault that their OS is broken.

**Comment by N5N3** [comment]
So the problem is that `fma` on windows isnot hardware based even if the cpu has that instruction?

**Comment by oscardssmith** [comment]
The problem is that LLVM will constant fold using the software version sometimes since it assumes that the libm version (provided by your OS) is correct. This leads to really weird bugs like the same `fma` call in a function producing different results depending on whether LLVM is able to constant fold based on context.

**Comment by KristofferC** [comment]
To me, completely disabling it on Windows feels way too heavy handed.

**Comment by yuyichao** [comment]
It should at least be safe to enable it if the compiling machine also has fma.

**Comment by oscardssmith** [comment]
I haven't tried it with windows built from source. Can someone with access to windows do so?

**Comment by N5N3** [comment]
I built master with 543ff67284ba05ea4d32d249ddcea045aaeb3249 using Cygwin64.
Unfortunately, test code in #43088 is verified broken.

**Comment by KristofferC** [comment]
Completely turning off fma on Windows is not an acceptable solution.

**Comment by vtjnash** [comment]
There is no functional or safe option until mingw64 is fixed

**Comment by oscardssmith** [comment]
honestly, we could let this be a functionality bug on windows rather than a performance bug.

**Comment by KristofferC** [comment]
> There is no functional or safe option until mingw64 is fixed

And still fma on Julia Windows has been used with little to no complaints for years. If the situation is the same on Windows as it has been so far, that is fine.

**Comment by ViralBShah** [comment]
I am assuming real code often will not be constant foldable - like you don't do `fma(1.0,1.0,1.0)` in real codes most likely. So, LLVM is likely going to generate the FMA instruction even on Windows in most cases. Even if you do end up in the libm version, the worst thing is it ends up being slow and not precise in some very special cases.

Would be nice if LLVM could be told to use our julia fma implementation, essentially showing it a Julia libm.

**Comment by N5N3** [comment]
Just notice that const propation is disabled for `sqrt`
https://github.com/JuliaLang/julia/blob/3a3cf028eb5e3138b0dfe5e31efce7f87f5cbfe3/base/compiler/optimize.jl#L547-L556.
Just wonder is that similar to `fma`? (LLVM uses a broken libm to fold constant)

**Comment by N5N3** [comment]
@oscardssmith Your test code in #43088 passed on `win10,master` if I add 069c0c4ac961cd975631a7d3ee060f40c5463e7d.

```julia
julia> versioninfo()
Julia Version 1.8.0-DEV.1190
Commit 069c0c4ac9 (2021-12-23 04:19 UTC)        
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-9600KF CPU @ 3.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, skylake)

julia> @code_llvm optimize = false fma(1.0,1.0,1.0)
;  @ floatfuncs.jl:418 within `fma`
; Function Attrs: uwtable
define double @julia_fma_819(double %0, double %1, double %2) #0 {
top:
  %3 = call {}*** @julia.get_pgcstack()
  %4 = bitcast {}*** %3 to {}**
  %current_task = getelementptr inbounds {}*, {}** %4, i64 -12
  %5 = bitcast {}** %current_task to i64*
  %world_age = getelementptr inbounds i64, i64* %5, i64 13
; ┌ @ floatfuncs.jl:411 within `fma_llvm`
   %6 = call double @llvm.fma.f64(double %0, double %1, double %2)
; └
  ret double %6
}

julia> @code_llvm fma(1.0,1.0,1.0)
;  @ floatfuncs.jl:418 within `fma`
; Function Attrs: uwtable
define double @julia_fma_847(double %0, double %1, double %2) #0 {
top:
; ┌ @ floatfuncs.jl:411 within `fma_llvm`
   %3 = call double @llvm.fma.f64(double %0, double %1, double %2)
; └
  ret double %3
}

julia> func=fma;

julia> using Test

julia> @test func(-1.9369631f13, 2.1513551f-7, -1.7354427f-24) == -4.1670958f6
Test Passed
  Expression: func(-1.9369631f13, 2.1513551f-7, -1.7354427f-24) == -4.1670958f6
   Evaluated: -4.1670958f6 == -4.1670958f6

julia> for func in (fma, fma)
           @test func(-1.9369631f13, 2.1513551f-7, -1.7354427f-24) == -4.1670958f6
       end

julia> f() = fma(-1.9369631f13, 2.1513551f-7, -1.7354427f-24)
f (generic function with 1 method)

julia> @code_llvm optimize = false f()
;  @ REPL[8]:1 within `f`
; Function Attrs: uwtable
define float @julia_f_943() #0 {
top:
  %0 = call {}*** @julia.get_pgcstack()
  %1 = bitcast {}*** %0 to {}**
  %current_task = getelementptr inbounds {}*, {}** %1, i64 -12
  %2 = bitcast {}** %current_task to i64*
  %world_age = getelementptr inbounds i64, i64* %2, i64 13
; ┌ @ floatfuncs.jl:417 within `fma`
; │┌ @ floatfuncs.jl:410 within `fma_llvm`
    %3 = call float @llvm.fma.f32(float 0xC2B19DD800000000, float 0x3E8CE00000000000, float 0xBB00C8BF80000000)
; └└
  ret float %3
}

julia> @code_llvm f()
;  @ REPL[8]:1 within `f`
; Function Attrs: uwtable
define float @julia_f_945() #0 {
top:
  ret float 0xC14FCADBE0000000
}

julia> @test f() == -4.1670958f6
Test Passed
  Expression: f() == -4.1670958f6
   Evaluated: -4.1670958f6 == -4.1670958f6
```

**Comment by oscardssmith** [comment]
Is it possible to check the OS at this point in the code? This looks like a good change, but I'd prefer if everyone's performance didn't regress because windows doesn't know how to program.

**Comment by N5N3** [comment]
Currently const propagtion of `f() = sin(sin(sin(sin(1))))` is blocked by `sqrt_llvm` because it is included in `is_pure_intrinsic_infer`. See [Discourse: Constant propagation depth](https://discourse.julialang.org/t/constant-propagation-depth/73012/7) 
Adding `fma_float` to `is_pure_intrinsic_infer` will also block const propagtion on Julia side.
I'm not sure how many math call use fma.
Maybe we can force use `fma_emulated` during inference and replace it with `fma_llvm` during `code_gen`?
(`sqrt` can be handled in a similar way?)

**Comment by oscardssmith** [comment]
Pretty much all math uses `fma` (at least the parts I've touched). That said, I have been pretty careful to use `muladd` where possible which sidesteps the issue. Forcing `fma_emulated` during inference would work, although you would have to be careful no to upset inlining thresholds. `fma` should always inline, while `fma_emulated` never should. The other possible solution would be to figure out why LLVM is constant folding with incorrect definitions and fix that.

**Comment by N5N3** [comment]
Well, just notice that
```julia
fma(x::Float32, y::Float32, z::Float32) = Core.Intrinsics.have_fma(Float32) ? fma_llvm(x,y,z) : fma_emulated(x,y,z)
```
Since `Core.Intrinsics.have_fma(Float32)` is not a const during inference, the const propagation for `f() = fma(1.0,1.0,1.0)` is blocked on julia side anyway like `sqrt`. Adding `fma_float` to `is_pure_intrinsic_infer` should cause no runtime regession on any system.

In this sense, we only need something like:
```julia
@static if Sys.iswindows()
    # Windows fma is broken currently. Always use `fma_llvm` to avoid runtime regression.
    # The result should be correct if your CPU has fma instruction.
    fma(x::Float32, y::Float32, z::Float32) = fma_llvm(x,y,z)
    fma(x::Float64, y::Float64, z::Float64) = fma_llvm(x,y,z)
else
    # Disable LLVM's fma if it is incorrect, e.g. because LLVM falls back
    # onto a broken system libm; if so, use a software emulated fma
    fma(x::Float32, y::Float32, z::Float32) = Core.Intrinsics.have_fma(Float32) ? fma_llvm(x,y,z) : fma_emulated(x,y,z)
    fma(x::Float64, y::Float64, z::Float64) = Core.Intrinsics.have_fma(Float64) ? fma_llvm(x,y,z) : fma_emulated(x,y,z)
end
```
